bigsea-controller/service/api/actuator/plugins/kvm_upv.py

```diff
@@ -0,0 +1,158 @@
+from service.api.actuator.actuator import Actuator
+from service.exceptions.kvm_exceptions import InstanceNotFoundException
+
+import ConfigParser
+import subprocess
+
+# TODO: documentation
+
+class KVM_Actuator_UPV(Actuator):
+
+    def __init__(self, conn, iops_reference, bs_reference):
+        self.conn = conn
+        self.config = ConfigParser.RawConfigParser()
+        self.config.read("controller.cfg")
+        self.one_user = self.config.get("actuator", "one_username")
+        self.one_password = self.config.get("actuator", "one_password")
+        self.one_url = self.config.get("actuator", "one_url")
+        self.iops_reference = iops_reference
+        self.bs_reference = bs_reference
+
+    # TODO: validation
+    def prepare_environment(self, vm_data):
+        self.adjust_resources(vm_data)
+
+
+    # TODO: validation
+    # This method receives as argument a map {vm-id:CPU cap}
+    def adjust_resources(self, vm_data):
+        instances_locations = {}
+
+        # Discover vm_id - compute nodes map
+        for instance in vm_data.keys():
+            # Access compute nodes to discover vms location
+            instances_locations[instance] = self._find_host(instance)
+
+        for instance in vm_data.keys():
+            # Access a compute node and change cap
+            self._change_vcpu_quota(instances_locations[instance],
+                                    instance, int(vm_data[instance]))
+
+            self._change_io_quota(instances_locations[instance],
+                                    instance, int(vm_data[instance]))
+
+
+    # TODO: validation
+    def get_allocated_resources(self, vm_id):
+        # Access compute nodes to discover vm location
+        host = self._find_host(vm_id)
+
+        # List all the vms to get the ONE id and map with the KVM id
+        virsh_list = "virsh list"
+        command = ("ssh root@niebla.i3m.upv.es \'ssh %s \"%s\"\'" % (host, virsh_list))
+        stdin, stdout, stderr = self.conn.exec_command(command)
+
+        vm_list = stdout.read().split("\n")
+        virsh_id = self._extract_id(vm_list, vm_id)
+        virsh_schedinfo = (("virsh schedinfo %s | grep vcpu_quota " % (virsh_id)) + 
+                                                             "| awk \'{print $3}\'")
+
+        command = ("ssh root@niebla.i3m.upv.es \'ssh %s \'%s\'\'" % (host, virsh_schedinfo))
```

Comment:
We could move this code to some parameter in the configuration like: compute-nodes-frontend

It is not good to use hacks like this but in this case is completely necessary, so we can minimize the ugliness of the code organizing better things like this one. :-)

Reply:
done